File: backend-legacy/src/services/tools/calculator.ts
/**
 * Calculator Tool — safely evaluates math expressions
 *
 * AI detects when the user asks to calculate something and runs it.
 * Uses a restricted math evaluator (no eval) for safety.
 */

import { registerTool } from './index';
import type { ToolDefinition, ToolExecutor } from './index';

// ─── Safe Math Evaluator ──────────────────────────────────
// A simple recursive-descent parser for arithmetic expressions.
// Supports: +, -, *, /, ^, %, parentheses, and common math functions.
// No arbitrary code execution — only math operations.

type Token =
  | { type: 'number'; value: number }
  | { type: 'op'; value: string }
  | { type: 'paren'; value: '(' | ')' }
  | { type: 'func'; value: string }
  | { type: 'comma'; value: ',' }
  | { type: 'ident'; value: string };

const MATH_FUNCS = new Set([
  'abs', 'floor', 'ceil', 'round', 'sqrt', 'cbrt',
  'sin', 'cos', 'tan', 'asin', 'acos', 'atan',
  'log', 'log2', 'log10', 'ln',
  'exp', 'pow', 'max', 'min',
  'pi', 'e',
]);

const MATH_CONSTANTS: Record<string, number> = {
  pi: Math.PI,
  e: Math.E,
};

function tokenize(expr: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  while (i < expr.length) {
    const ch = expr[i];
    if (/\s/.test(ch)) { i++; continue; }

    // Number
    if (/\d/.test(ch)) {
      let num = '';
      while (i < expr.length && /[\d.]/.test(expr[i])) { num += expr[i]; i++; }
      tokens.push({ type: 'number', value: parseFloat(num) });
      continue;
    }

    // Function name or identifier (like pi, e)
    if (/[a-zA-Z_]/.test(ch)) {
      let name = '';
      while (i < expr.length && /[a-zA-Z0-9_]/.test(expr[i])) { name += expr[i]; i++; }
      if (name in MATH_CONSTANTS) {
        tokens.push({ type: 'number', value: MATH_CONSTANTS[name] });
      } else if (MATH_FUNCS.has(name)) {
        tokens.push({ type: 'func', value: name });
      } else {
        tokens.push({ type: 'ident', value: name });
      }
      continue;
    }

    // Operators
    if ('+-*/^%'.includes(ch)) {
      tokens.push({ type: 'op', value: ch });
      i++;
      continue;
    }

    // Parentheses
    if (ch === '(' || ch === ')') {
      tokens.push({ type: 'paren', value: ch });
      i++;
      continue;
    }

    // Comma
    if (ch === ',') {
      tokens.push({ type: 'comma', value: ',' });
      i++;
      continue;
    }

    // Unknown char — skip
    i++;
  }
  return tokens;
}

let pos = 0;
let tokens: Token[] = [];

function peek(): Token | null { return pos < tokens.length ? tokens[pos] : null; }
function consume(): Token | null { return pos < tokens.length ? tokens[pos++] : null; }

function parsePrimary(): number {
  const tok = peek();
  if (!tok) throw new Error('Unexpected end of expression');

  if (tok.type === 'number') {
    consume();
    return tok.value;
  }

  if (tok.type === 'paren' && tok.value === '(') {
    consume(); // eat '('
    const val = parseExpression();
    const close = consume();
    if (!close || close.type !== 'paren' || close.value !== ')') {
      throw new Error('Missing closing parenthesis');
    }
    return val;
  }

  if (tok.type === 'func') {
    consume();
    // Parse function arguments (inside parentheses)
    if (peek()?.type === 'paren' && peek()?.value === '(') {
      consume(); // eat '('
      const args: number[] = [];
      while (!(peek()?.type === 'paren' && peek()?.value === ')')) {
        args.push(parseExpression());
        if (peek()?.type === 'comma') consume();
      }
      consume(); // eat ')'
      return applyFunction(tok.value, args);
    }
    // No parentheses — might be a constant-like usage
    return applyFunction(tok.value, []);
  }

  if (tok.type === 'op' && tok.value === '-') {
    consume(); // unary minus
    return -parsePrimary();
  }
  if (tok.type === 'op' && tok.value === '+') {
    consume(); // unary plus
    return parsePrimary();
  }

  throw new Error(`Unexpected token: ${JSON.stringify(tok)}`);
}

function applyFunction(name: string, args: number[]): number {
  switch (name) {
    case 'abs': return Math.abs(args[0]);
    case 'floor': return Math.floor(args[0]);
    case 'ceil': return Math.ceil(args[0]);
    case 'round': return Math.round(args[0]);
    case 'sqrt': return Math.sqrt(args[0]);
    case 'cbrt': return Math.cbrt(args[0]);
    case 'sin': return Math.sin(args[0]);
    case 'cos': return Math.cos(args[0]);
    case 'tan': return Math.tan(args[0]);
    case 'asin': return Math.asin(args[0]);
    case 'acos': return Math.acos(args[0]);
    case 'atan': return Math.atan(args[0]);
    case 'log':
    case 'log10': return Math.log10(args[0]);
    case 'log2': return Math.log2(args[0]);
    case 'ln': return Math.log(args[0]);
    case 'exp': return Math.exp(args[0]);
    case 'pow': return Math.pow(args[0], args[1] || 1);
    case 'max': return Math.max(...args);
    case 'min': return Math.min(...args);
    default: throw new Error(`Unknown function: ${name}`);
  }
}

function parseExponent(): number {
  let left = parsePrimary();
  while (peek()?.type === 'op' && peek()?.value === '^') {
    consume();
    const right = parseExponent();
    left = Math.pow(left, right);
  }
  return left;
}

function parseMultiplicative(): number {
  let left = parseExponent();
  while (peek()?.type === 'op' && ('*/%'.includes(peek()?.value || ''))) {
    const op = consume()!.value;
    const right = parseExponent();
    if (op === '*') left *= right;
    else if (op === '/') {
      if (right === 0) throw new Error('Division by zero');
      left /= right;
    }
    else if (op === '%') {
      if (right === 0) throw new Error('Division by zero (modulo)');
      left %= right;
    }
  }
  return left;
}

function parseAdditive(): number {
  let left = parseMultiplicative();
  while (peek()?.type === 'op' && (peek()?.value === '+' || peek()?.value === '-')) {
    const op = consume()!.value;
    const right = parseMultiplicative();
    if (op === '+') left += right;
    else left -= right;
  }
  return left;
}

function parseExpression(): number {
  return parseAdditive();
}

function safeEval(expr: string): number {
  tokens = tokenize(expr);
  pos = 0;
  const result = parseExpression();
  if (peek() !== null) {
    throw new Error(`Unexpected tokens after expression`);
  }
  return result;
}

// ─── Tool Definition ──────────────────────────────────────

const definition: ToolDefinition = {
  id: 'calculator',
  name: 'Calculator',
  description: 'Evaluate math expressions: arithmetic, trigonometry, logarithms, powers, and more',
  version: '1.0.0',
  icon: '🧮',
  params: [
    { name: 'expression', type: 'string', description: 'Math expression to evaluate (e.g., "2 + 2", "sqrt(144)", "sin(45)")', required: true },
  ],
};

const execute: ToolExecutor = async (params) => {
  const expr = String(params.expression || params.query || '');
  if (!expr.trim()) {
    return { success: false, output: 'Please provide a math expression to evaluate.' };
  }

  try {
    const result = safeEval(expr);
    const formatted = Number.isInteger(result) ? result.toString() : result.toFixed(4);
    return {
      success: true,
      output: `${expr} = ${formatted}`,
      data: { expression: expr, result },
    };
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    return {
      success: false,
      output: `Could not evaluate "${expr}": ${msg}`,
    };
  }
};

// Auto-detect calculator intent from user input
const CALC_PATTERNS = [
  /(\d+)\s*([+\-*/^])\s*(\d+)/,
  /calculate\s+(.+)/i,
  /what (is|'s)\s+(.+)/,
  /compute\s+(.+)/i,
  /solve\s+(.+)/i,
];

function detect(input: string): { confidence: number; params: Record<string, unknown> } | null {
  const lower = input.toLowerCase();

  // Skip if it's clearly a conversion
  if (lower.includes('convert') || lower.includes('cm') || lower.includes('inches') || lower.includes('feet')) {
    return null;
  }

  // Contains clear math operators
  if (/[\d]\s*[+\-*/^%]\s*[\d]/.test(input) && !lower.includes('http')) {
    return { confidence: 0.8, params: { expression: input.trim() } };
  }

  // Starts with "calculate", "what is", "compute", "solve"
  for (const pat of CALC_PATTERNS) {
    const m = input.match(pat);
    if (m) {
      const expr = m[m.length - 1].trim();
      if (expr.length > 0 && expr.length < 200) {
        return { confidence: 0.7, params: { expression: expr } };
      }
    }
  }

  return null;
}

// ─── Register ─────────────────────────────────────────────

export function registerCalculatorTool(): void {
  registerTool(definition, execute);
  console.log('[tools] Calculator registered with safe math evaluator');
}

File: backend/app/tools/calculator.py
"""Calculator Tool — mirrors backend/src/services/tools/calculator.ts.

Safely evaluates math expressions with a recursive-descent parser (no eval).
"""

from __future__ import annotations

import math
import re
from typing import Any, Callable

from ..logger import info as log_info
from . import ToolContext, ToolDefinition, ToolParam, ToolResult, register_tool

MATH_FUNCS = {
    "abs", "floor", "ceil", "round", "sqrt", "cbrt",
    "sin", "cos", "tan", "asin", "acos", "atan",
    "log", "log2", "log10", "ln", "exp", "pow", "max", "min", "pi", "e",
}

MATH_CONSTANTS: dict[str, float] = {"pi": math.pi, "e": math.e}


class _EvalState:
    def __init__(self) -> None:
        self.tokens: list[tuple[str, Any]] = []
        self.pos = 0

    def peek(self) -> tuple[str, Any] | None:
        return self.tokens[self.pos] if self.pos < len(self.tokens) else None

    def consume(self) -> tuple[str, Any] | None:
        tok = self.peek()
        if tok is not None:
            self.pos += 1
        return tok


def _tokenize(expr: str) -> list[tuple[str, Any]]:
    tokens: list[tuple[str, Any]] = []
    i = 0
    while i < len(expr):
        ch = expr[i]
        if ch.isspace():
            i += 1
            continue
        if ch.isdigit():
            num = ""
            while i < len(expr) and (expr[i].isdigit() or expr[i] == "."):
                num += expr[i]
                i += 1
            tokens.append(("number", float(num)))
            continue
        if ch.isalpha() or ch == "_":
            name = ""
            while i < len(expr) and (expr[i].isalnum() or expr[i] == "_"):
                name += expr[i]
                i += 1
            if name.lower() in MATH_CONSTANTS:
                tokens.append(("number", MATH_CONSTANTS[name.lower()]))
            elif name.lower() in MATH_FUNCS:
                tokens.append(("func", name.lower()))
            else:
                tokens.append(("ident", name))
            continue
        if ch in "+-*/^%":
            tokens.append(("op", ch))
            i += 1
            continue
        if ch in "()":
            tokens.append(("paren", ch))
            i += 1
            continue
        if ch == ",":
            tokens.append(("comma", ch))
            i += 1
            continue
        i += 1  # Unknown char — skip
    return tokens


def _apply_function(name: str, args: list[float]) -> float:
    if name == "abs":
        return abs(args[0])
    if name == "floor":
        return math.floor(args[0])
    if name == "ceil":
        return math.ceil(args[0])
    if name == "round":
        return round(args[0])
    if name == "sqrt":
        return math.sqrt(args[0])
    if name == "cbrt":
        return math.copysign(abs(args[0]) ** (1 / 3), args[0])
    if name == "sin":
        return math.sin(args[0])
    if name == "cos":
        return math.cos(args[0])
    if name == "tan":
        return math.tan(args[0])
    if name == "asin":
        return math.asin(args[0])
    if name == "acos":
        return math.acos(args[0])
    if name == "atan":
        return math.atan(args[0])
    if name in ("log", "log10"):
        return math.log10(args[0])
    if name == "log2":
        return math.log2(args[0])
    if name == "ln":
        return math.log(args[0])
    if name == "exp":
        return math.exp(args[0])
    if name == "pow":
        return math.pow(args[0], args[1] if len(args) > 1 else 1)
    if name == "max":
        return max(args)
    if name == "min":
        return min(args)
    raise ValueError(f"Unknown function: {name}")


class _Parser:
    def __init__(self, tokens: list[tuple[str, Any]]) -> None:
        self.st = _EvalState()
        self.st.tokens = tokens

    def parse_expression(self) -> float:
        return self._additive()

    def _primary(self) -> float:
        tok = self.st.peek()
        if tok is None:
            raise ValueError("Unexpected end of expression")
        kind, value = tok
        if kind == "number":
            self.st.consume()
            return value
        if kind == "paren" and value == "(":
            self.st.consume()
            val = self.parse_expression()
            close = self.st.consume()
            if close is None or close[0] != "paren" or close[1] != ")":
                raise ValueError("Missing closing parenthesis")
            return val
        if kind == "func":
            self.st.consume()
            nxt = self.st.peek()
            if nxt and nxt[0] == "paren" and nxt[1] == "(":
                self.st.consume()
                args: list[float] = []
                while not (self.st.peek() and self.st.peek()[0] == "paren" and self.st.peek()[1] == ")"):
                    if self.st.peek() is None:
                        raise ValueError("Missing closing parenthesis")
                    args.append(self.parse_expression())
                    if self.st.peek() and self.st.peek()[0] == "comma":
                        self.st.consume()
                self.st.consume()  # eat ')'
                return _apply_function(value, args)
            return _apply_function(value, [])
        if kind == "op" and value == "-":
            self.st.consume()
            return -self._primary()
        if kind == "op" and value == "+":
            self.st.consume()
            return self._primary()
        raise ValueError(f"Unexpected token: {tok}")

    def _exponent(self) -> float:
        left = self._primary()
        while self.st.peek() and self.st.peek()[0] == "op" and self.st.peek()[1] == "^":
            self.st.consume()
            right = self._exponent()
            left = math.pow(left, right)
        return left

    def _multiplicative(self) -> float:
        left = self._exponent()
        while self.st.peek() and self.st.peek()[0] == "op" and self.st.peek()[1] in "*/%":
            op = self.st.consume()[1]
            right = self._exponent()
            if op == "*":
                left *= right
            elif op == "/":
                if right == 0:
                    raise ValueError("Division by zero")
                left /= right
            else:
                if right == 0:
                    raise ValueError("Division by zero (modulo)")
                left %= right
        return left

    def _additive(self) -> float:
        left = self._multiplicative()
        while self.st.peek() and self.st.peek()[0] == "op" and self.st.peek()[1] in "+-":
            op = self.st.consume()[1]
            right = self._multiplicative()
            left = left + right if op == "+" else left - right
        return left


def safe_eval(expr: str) -> float:
    parser = _Parser(_tokenize(expr))
    result = parser.parse_expression()
    if parser.st.peek() is not None:
        raise ValueError("Unexpected tokens after expression")
    return result


DEFINITION = ToolDefinition(
    id="calculator",
    name="Calculator",
    description="Evaluate math expressions: arithmetic, trigonometry, logarithms, powers, and more",
    version="1.0.0",
    icon="\U0001f9ee",
    params=[
        ToolParam(
            name="expression",
            type="string",
            description='Math expression to evaluate (e.g., "2 + 2", "sqrt(144)", "sin(45)")',
            required=True,
        ),
    ],
)


async def execute(params: dict[str, Any], ctx: ToolContext) -> ToolResult:
    expr = str(params.get("expression") or params.get("query") or "")
    if not expr.strip():
        return ToolResult(success=False, output="Please provide a math expression to evaluate.")
    try:
        result = safe_eval(expr)
        formatted = str(int(result)) if float(result).is_integer() else f"{result:.4f}"
        return ToolResult(
            success=True,
            output=f"{expr} = {formatted}",
            data={"expression": expr, "result": result},
        )
    except Exception as e:  # noqa: BLE001
        return ToolResult(success=False, output=f'Could not evaluate "{expr}": {e}')


CALC_PATTERNS = [
    re.compile(r"(\d+)\s*([+\-*/^])\s*(\d+)"),
    re.compile(r"calculate\s+(.+)", re.IGNORECASE),
    re.compile(r"what (is|'s)\s+(.+)"),
    re.compile(r"compute\s+(.+)", re.IGNORECASE),
    re.compile(r"solve\s+(.+)", re.IGNORECASE),
]


def detect(input_text: str) -> dict[str, Any] | None:
    """Auto-detect calculator intent from user input."""
    lower = input_text.lower()
    if any(k in lower for k in ("convert", "cm", "inches", "feet")):
        return None
    if re.search(r"[\d]\s*[+\-*/^%]\s*[\d]", input_text) and "http" not in lower:
        return {"confidence": 0.8, "params": {"expression": input_text.strip()}}
    for pat in CALC_PATTERNS:
        m = pat.search(input_text)
        if m:
            expr = m.group(m.lastindex or 1).strip() if m.groups() else ""
            if not expr:
                continue
            if 0 < len(expr) < 200:
                return {"confidence": 0.7, "params": {"expression": expr}}
    return None


def register_calculator_tool() -> None:
    register_tool(DEFINITION, execute)
    log_info("[tools] Calculator registered with safe math evaluator")
